File: spring-webflux/src/main/java/software/darkmatter/school/blog/config/security/CustomAuthenticationConverter.java
package software.darkmatter.school.blog.config.security;

import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.web.server.authentication.ServerAuthenticationConverter;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ServerWebExchange;
import reactor.core.publisher.Mono;
import software.darkmatter.school.blog.domain.user.business.UserService;

import java.util.Optional;

@Component
@RequiredArgsConstructor
public class CustomAuthenticationConverter implements ServerAuthenticationConverter {

    public static final String PREFIX_USER_ID = "User-Id ";
    public static final String AUTHORIZATION = "Authorization";

    private final UserService userService;

    @Override
    public Mono<Authentication> convert(ServerWebExchange exchange) {
        Optional<Mono<Authentication>> authenticationMono =
            Optional.ofNullable(exchange.getRequest().getHeaders().get(AUTHORIZATION))
                    .flatMap(headers -> headers.stream().findFirst())
                    .filter(header -> header.startsWith(PREFIX_USER_ID))
                    .map(authorizationHeader -> {
                        String userId = authorizationHeader.substring(PREFIX_USER_ID.length());
                        return userService.getById(Long.valueOf(userId))
                                          .map(
                                              user -> new SimpleAuthentication(
                                                  user.getId(),
                                                  user.getFirstName()
                                              )
                                          );
                    });

        return authenticationMono.orElse(Mono.empty());
    }
}

File: kotlin-spring-webflux/src/main/kotlin/software/darkmatter/school/blog/config/security/CustomAuthenticationConverter.kt
package software.darkmatter.school.blog.config.security

import kotlinx.coroutines.reactor.mono
import org.springframework.security.core.Authentication
import org.springframework.security.web.server.authentication.ServerAuthenticationConverter
import org.springframework.stereotype.Component
import org.springframework.web.server.ServerWebExchange
import reactor.core.publisher.Mono
import software.darkmatter.school.blog.domain.user.business.UserService

@Component
class CustomAuthenticationConverter(
    private val userService: UserService,
) : ServerAuthenticationConverter {

    override fun convert(exchange: ServerWebExchange): Mono<Authentication> =
        exchange.request.headers[AUTHORIZATION]
            ?.first { it.startsWith(PREFIX_USER_ID) }
            ?.let { authorizationHeader: String ->
                val userId = authorizationHeader.substring(PREFIX_USER_ID.length)

                mono {
                    val user = userService.getById(userId.toLong())

                    SimpleAuthentication(user.id!!, user.firstName)
                }
            } ?: Mono.empty()

    companion object {
        const val PREFIX_USER_ID = "User-Id "
        const val AUTHORIZATION = "Authorization"
    }
}
